lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py

```diff
@@ -814,406 +358,136 @@ def _aggregate_grads_async(self, tag: str) -> None:
-        # # the list of lists into a numpy array, get the avg
-        # agg_staleness_arr = np.hstack(self._aggregator_staleness_track_rounds)
-        # logger.info(f"==== aggregator avg staleness:
-        # {np.mean(agg_staleness_arr)}")
-
-        # # per trainer analytics
-        # if self._round % 100 == 0: for k, v in
-        #     self._per_trainer_staleness_track.items(): trainer_staleness_arr =
-        #         np.array(v) logger.info( f"Trainer {k} staleness info. Min
-        #         {np.min(trainer_staleness_arr)}, " f"Max
-        #             {np.max(trainer_staleness_arr)}, " f"Avg
-        #             {np.mean(trainer_staleness_arr)}, " f"P50
-        #             {np.median(trainer_staleness_arr)}, " f"P90
-        #             {np.percentile(trainer_staleness_arr, 90)}, " f"P99
-        #             {np.percentile(trainer_staleness_arr, 99)}" )
-
-        # total_training_time_all_trainers = 0 for k, v in
-        # self._track_trainer_version_duration_s.items():
-        #     total_training_time_all_trainers += v["total_training_time_s"]
-        # avg_training_time = total_training_time_all_trainers / len(
-        #     self._track_trainer_version_duration_s ) logger.info( f"Avg
-        # training time {avg_training_time} across "
-        # f"{len(self._track_trainer_version_duration_s)} trainers" )
-        if self.var_good_enough:
-            channel._selector.remove_from_selected_ends(channel._ends, end)
-        logger.debug("Agg goal reached, so resetting trainer end states in the channel")
-        channel.cleanup_recvd_ends()
 
-    def _aggregate_grads_sync(self, tag: str) -> None:
-        """Aggregate trainer gradients synchronously."""
-        logger.info("starting aggregate_grads_sync")
-        self.log_memory("start _aggregate_grads_sync", self.device)
-        self.print_trainable_params_stats(location="[start,_aggregate_grads_sync()]")
-        if self.ends_not_selected_yet:
-            logger.info("no ends selected yet")
-            return
-
-        channel = self.cm.get_by_tag(tag)
-        if not channel:
-            return
-
-        logger.debug(f"Channel {channel} found for tag {tag}")
-        # receive local model parameters from a trainer who arrives first NOTE:
-        # (DG) Right now, the leave notifications also cause a message to be
-        # processed and yield (None,None) from recv_fifo().
-        if channel.ends(VAL_CH_STATE_RECV) is None:
-            logger.info("no ends yet")
-            return
-
-        total = 0
+            self._updates_in_queue -= self._agg_goal
+            self._agg_goal_cnt = 0 # Reset for the next batch
+            # ASYNC: Clean up ends that just sent data
+            logger.debug("Agg goal reached, so resetting trainer end states in the channel")
+            channel.cleanup_recvd_ends()
 
-        # receive local model parameters from trainers
+    def aggregate_and_collect(self, tag, channel):
```

Comment:
Name of this function and docstring is a bit confusing. I almost thought that this is _aggregate_grads_sync()